ra2ce/network/network_wrappers/osm_network_wrapper/osm_utils.py

```diff
@@ -93,3 +96,151 @@ def from_shapefile_to_poly(shapefile: Path, out_path: Path, outname: str = ""):
+
+    # the nearest edge minimizes the non-zero distance to the node
+    (u, v, data, key, geom), dist = min(edge_distances, key=lambda x: x[1] if x[1] > 0 else float('inf'))
+    utils.log(f"Found nearest edge ({u, v, data, key}) to node {node}")
+
+    # return results requested by caller
+    if data == 0 and key == 0:
+        if return_dist and return_geom:
+            return {
+                "node": node,
+                "nearest_edge": (u, v, geom, dist)}
+        elif return_dist:
+            return {
+                "node": node,
+                "nearest_edge": (u, v, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v)}
+    elif data == 0:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, key, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, key)}
+    elif key == 0:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, data)}
+    else:
+        if return_dist and return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, geom, dist)}
+        elif return_dist:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, dist)}
+        elif return_geom:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key, geom)}
+        else:
+            return {"node": node,
+                    "nearest_edge": (u, v, data, key)}
+
+
+def _is_endnode_simplified(graph: MultiDiGraph, node: int) -> bool:
+    """
+    Based on osmnx. osmnx rules 3 and 4 are removed. Hence, the name _is_endpoint_simplified.
+    """
+    neighbors = set(list(graph.predecessors(node)) + list(graph.successors(node)))
+    n = len(neighbors)
+    d = graph.degree(node)
+
+    # rule 1
+    if node in neighbors:
+        # if the node appears in its list of neighbors, it self-loops
+        # this is always an endpoint.
+        return True
+
+    # rule 2
+    elif graph.out_degree(node) == 0 or graph.in_degree(node) == 0:
+        # if node has no incoming edges or no outgoing edges, it is an endpoint
+        return True
+
```

Comment:
What do you do if none of the above conditions is met? Do not forget to add a default return statement.